Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabPublicHandler.cpp

```diff
@@ -726,20 +731,23 @@ namespace AzToolsFramework
             }
 
             PrefabDom beforeState;
+            bool isBeforeStateValid = m_instanceDomGeneratorInterface->GenerateEntityDom(beforeState, *entity);
             AZ::EntityId beforeParentId;
-            m_prefabUndoCache.Retrieve(entityId, beforeState, beforeParentId);
+            m_prefabUndoCache.Retrieve(entityId, beforeParentId);
 
             PrefabDom afterState;
             AZ::EntityId afterParentId;
             AZ::TransformBus::EventResult(afterParentId, entityId, &AZ::TransformBus::Events::GetParentId);
-
-            m_instanceToTemplateInterface->GenerateDomForEntity(afterState, *entity);
+            bool isAfterStateValid = m_instanceToTemplateInterface->GenerateDomForEntity(afterState, *entity);
 
             PrefabDom patch;
-            m_instanceToTemplateInterface->GeneratePatch(patch, beforeState, afterState);
-            m_instanceToTemplateInterface->AppendEntityAliasToPatchPaths(patch, entityId);
+            if (isBeforeStateValid && isAfterStateValid)
+            {
+                m_instanceToTemplateInterface->GeneratePatch(patch, beforeState, afterState);
+                m_instanceToTemplateInterface->AppendEntityAliasToPatchPaths(patch, entityId);
+            }
 
-            if (patch.IsArray() && !patch.Empty() && beforeState.IsObject())
+            if (patch.IsArray() && !patch.Empty())
```

Comment:
The 'beforeState.IsObject()' is still a good check to keep imo because we don't know what might get put into the 'afterState' dom. Is there a reason for removing it?

Reply:
good catch! I thought `isBeforeStateValid` could capture this but it actually does not.

```cpp
if (isBeforeStateValid && isAfterStateValid && beforeState.IsObject() && afterState.IsObject())
{
    m_instanceToTemplateInterface->GeneratePatch(patch, beforeState, afterState);
    m_instanceToTemplateInterface->AppendEntityAliasToPatchPaths(patch, entityId);
}
```

Reply:
I suggest reverting my changes for DOM check in this PR and address these changes in another PR.